### PR TITLE
Fix Attach/Detach NC from Network Calls

### PR DESCRIPTION
**Pull request by jaer-tsun**
**Reason for Change**:
CNS was not returning the correct response when invoking CNI.
Attach/Detach calls were not limited to Batch orchestrator type.

**Special Note**:
- Created VM with CNS and DNC
- SWIFT setup
- Created docker containers
- Verified networking and responses before/after attach/detach calls

### CODE REVIEW

File: cns/networkcontainers/networkcontainers.go [hunk 1]
```diff
@@ -121,6 +125,9 @@ func getNetworkConfig(configFilePath string) ([]byte, error) {
 	flatNetConfigMap[versionStr] = configMap[versionStr].(string)
 	flatNetConfigMap[nameStr] = configMap[nameStr].(string)
 
+	// insert CNS URL
+	flatNetConfigMap[cnsURL] = strings.Replace(cnsUrl, "tcp://", "http://", -1)
```
Comment (tamilmani1989) [inline cns/networkcontainers/networkcontainers.go]: if cnsURL is empty.. flatNetConfigMap[cnsURL] will this work?

File: cns/networkcontainers/networkcontainers_windows.go [hunk 2]
```diff
@@ -188,7 +188,7 @@ func configureNetworkContainerNetworking(operation, podName, podNamespace, docke
 			{k8sPodNameStr, podName}}}
 	log.Printf("[Azure CNS] run time conf info %+v", cniRtConf)
 
-	netConfig, err := getNetworkConfig(netPluginConfig.networkConfigPath)
+	netConfig, err := getNetworkConfig(netPluginConfig.networkConfigPath, netPluginConfig.cnsUrl, true)
```
Comment (tamilmani1989) [inline cns/networkcontainers/networkcontainers_windows.go]: can we check orchestrator type and set to true here..by default in multitenancy case its false

Comment (jaer-tsun) [reply cns/networkcontainers/networkcontainers_windows.go]: I will check in restserver.go

File: cns/restserver/restserver.go [hunk 3]
```diff
@@ -1363,20 +1364,28 @@ func (service *HTTPRestService) attachNetworkContainerToNetwork(w http.ResponseW
...
-						returnMessage = fmt.Sprintf("[Azure CNS] Error. AttachContainerToNetwork failed %+v", err.Error())
+				switch service.state.OrchestratorType {
+				case cns.Batch:
+					var podInfo cns.KubernetesPodInfo
+					err := json.Unmarshal(existing.CreateNetworkContainerRequest.OrchestratorContext, &podInfo)
+					if err != nil {
 						returnCode = UnexpectedError
+						returnMessage = fmt.Sprintf("Unmarshalling orchestrator context failed with error %+v", err)
+					} else {
+						nc := service.networkContainer
```
Comment (tamilmani1989) [inline cns/restserver/restserver.go]: move this code to a function and reuse it in detach call also..looks repetitive

File: cns/networkcontainers/networkcontainers.go [hunk 4]
```diff
@@ -121,6 +127,12 @@ func getNetworkConfig(configFilePath string) ([]byte, error) {
 	flatNetConfigMap[versionStr] = configMap[versionStr].(string)
 	flatNetConfigMap[nameStr] = configMap[nameStr].(string)
 
+	// insert CNS URL
+	flatNetConfigMap[cnsURLStr] = strings.Replace(cnsURL, "tcp://", "http://", -1)
+
+	// enable exact match for pod name
```
Comment (ninzavivek) [inline cns/networkcontainers/networkcontainers.go]: this comment is not right.

File: cns/networkcontainers/networkcontainers.go [hunk 5]
```diff
@@ -121,6 +127,12 @@ func getNetworkConfig(configFilePath string) ([]byte, error) {
 	flatNetConfigMap[versionStr] = configMap[versionStr].(string)
 	flatNetConfigMap[nameStr] = configMap[nameStr].(string)
 
+	// insert CNS URL
```
Comment (ninzavivek) [inline cns/networkcontainers/networkcontainers.go]: what is the motivation of this change?

File: cns/networkcontainers/networkcontainers.go [hunk 6]
```diff
@@ -152,6 +153,11 @@ func pluginErr(err error, output []byte) error {
 
 			return &emsg
 		}
+	} else if len(output) > 0 {
```
Comment (tamilmani1989) [inline cns/networkcontainers/networkcontainers.go]: add comments here

File: cns/restserver/restserver.go [hunk 7]
```diff
@@ -1363,20 +1361,25 @@ func (service *HTTPRestService) attachNetworkContainerToNetwork(w http.ResponseW
...
+						netPluginConfig := service.getNetPluginConfiguration()
+						if err = nc.Attach(podInfo.PodName, podInfo.PodNamespace, req.Containerid, netPluginConfig); err != nil {
```
Comment (tamilmani1989) [inline cns/restserver/restserver.go]: just send podinfo

File: cns/restserver/restserver.go [hunk 8]
```diff
@@ -1426,20 +1429,24 @@ func (service *HTTPRestService) detachNetworkContainerFromNetwork(w http.Respons
...
+				case cns.Batch:
+					var podInfo cns.KubernetesPodInfo
+					err := json.Unmarshal(existing.CreateNetworkContainerRequest.OrchestratorContext, &podInfo)
+					if err != nil {
 						returnCode = UnexpectedError
+						returnMessage = fmt.Sprintf("Unmarshalling orchestrator context failed with error %+v", err)
+					} else {
+						nc := service.networkContainer
+						netPluginConfig := service.getNetPluginConfiguration()
+						if err = nc.Detach(podInfo.PodName, podInfo.PodNamespace, req.Containerid, netPluginConfig); err != nil {
```
Comment (tamilmani1989) [inline cns/restserver/restserver.go]: change here also

File: cns/networkcontainers/networkcontainers.go (hunk 6 above)
Comment (jaer-tsun) [reply cns/networkcontainers/networkcontainers.go]: done

File: cns/restserver/restserver.go [hunk 9]
```diff
@@ -1458,3 +1465,9 @@ func (service *HTTPRestService) detachNetworkContainerFromNetwork(w http.Respons
 	err = service.Listener.Encode(w, &detachResp)
 	log.Response(service.Name, detachResp, resp.ReturnCode, ReturnCodeToString(resp.ReturnCode), err)
 }
+
+func (service *HTTPRestService) getNetPluginConfiguration() *networkcontainers.NetPluginConfiguration {
```
Comment (tamilmani1989) [inline cns/restserver/restserver.go]: can we name this as getCNIPluginDetails

Comment (jaer-tsun) [reply cns/restserver/restserver.go]: done

File: cns/restserver/restserver.go [hunk 10]
```diff
@@ -1426,20 +1429,24 @@ func (service *HTTPRestService) detachNetworkContainerFromNetwork(w http.Respons
...
-					nc := service.networkContainer
-					pluginBinPath, _ := service.GetOption(acn.OptCNIPath).(string)
-					configPath, _ := service.GetOption(acn.OptCNIConfigFile).(string)
-					netPluginConfig := networkcontainers.NewNetPluginConfiguration(pluginBinPath, configPath)
-					if err = nc.Detach(podInfo.PodName, podInfo.PodNamespace, req.Containerid, netPluginConfig); err != nil {
-						returnMessage = fmt.Sprintf("[Azure CNS] Error. detachNetworkContainerFromNetwork failed %+v", err.Error())
+				switch service.state.OrchestratorType {
+				case cns.Batch:
+					var podInfo cns.KubernetesPodInfo
+					err := json.Unmarshal(existing.CreateNetworkContainerRequest.OrchestratorContext, &podInfo)
```
Comment (ninzavivek) [inline cns/restserver/restserver.go]: the rest call handlers are becoming too complex from readability point of view. Could you try encapsulating this logic in a new function?

File: cns/networkcontainers/networkcontainers.go [hunk 11]
```diff
@@ -171,25 +178,24 @@ func execPlugin(rt *libcni.RuntimeConf, netconf []byte, operation, path string)
...
-func (cn *NetworkContainers) Attach(podName, podNamespace, dockerContainerid string, netPluginConfig *NetPluginConfiguration) error {
+func (cn *NetworkContainers) Attach(podInfo cns.KubernetesPodInfo, dockerContainerid string, netPluginConfig *NetPluginConfiguration) error {
 	log.Printf("[Azure CNS] NetworkContainers.Attach called")
-	err := configureNetworkContainerNetworking(cniAdd, podName, podNamespace, dockerContainerid, netPluginConfig)
+	err := configureNetworkContainerNetworking(cniAdd, podInfo.PodName, podInfo.PodNamespace, dockerContainerid, netPluginConfig)
 	log.Printf("[Azure CNS] NetworkContainers.Attach finished")
 	return err
 }
 
 // Detach - attaches network container to network.
```
Comment (ninzavivek) [inline cns/networkcontainers/networkcontainers.go]: detaches?

Comment (jaer-tsun) [reply cns/networkcontainers/networkcontainers.go]: fixed

File: cns/restserver/restserver.go [hunk 12]
```diff
@@ -1337,60 +1335,13 @@ func (service *HTTPRestService) attachNetworkContainerToNetwork(w http.ResponseW
...
-			returnCode = InvalidParameter
-		}
-	}
-
-	resp := cns.Response{
-		ReturnCode: returnCode,
-		Message:    returnMessage,
-	}
-
+	resp := service.attachOrDetachHelper(req, "Attach", r.Method)
```
Comment (tamilmani1989) [inline cns/restserver/restserver.go]: move "Attach" to constant

File: cns/restserver/restserver.go [hunk 13]
```diff
@@ -1400,61 +1351,80 @@ func (service *HTTPRestService) detachNetworkContainerFromNetwork(w http.Respons
...
-		Message:    returnMessage,
-	}
+		Message:    returnMessage}
+}
 
-	detachResp := &cns.DetachContainerFromNetworkResponse{Response: resp}
-	err = service.Listener.Encode(w, &detachResp)
-	log.Response(service.Name, detachResp, resp.ReturnCode, ReturnCodeToString(resp.ReturnCode), err)
+func (service *HTTPRestService) getCNIPluginDetails() *networkcontainers.NetPluginConfiguration {
+	pluginBinPath, _ := service.GetOption(acn.OptCNIPath).(string)
```
Comment (tamilmani1989) [inline cns/restserver/restserver.go]: can you change this to OptNetpluginPath..my bad I asked you to change function name intead of this. also change configpath similarly..let the function name be getNetPluginDetails

File: cns/restserver/restserver.go [hunk 14]
```diff
@@ -1400,61 +1351,80 @@ func (service *HTTPRestService) detachNetworkContainerFromNetwork(w http.Respons
...
+		var podInfo cns.KubernetesPodInfo
+		err := json.Unmarshal(existing.CreateNetworkContainerRequest.OrchestratorContext, &podInfo)
+		if err != nil {
+			returnCode = UnexpectedError
+			returnMessage = fmt.Sprintf("Unmarshalling orchestrator context failed with error %+v", err)
+		} else {
+			nc := service.networkContainer
+			netPluginConfig := service.getCNIPluginDetails()
+			switch operation {
+			case "Attach":
```
Comment (tamilmani1989) [inline cns/restserver/restserver.go]: same here..move attach, detach to constants

File: cni/network/network_windows.go [hunk 15]
```diff
@@ -111,7 +111,7 @@ func updateSubnetPrefix(cnsNwConfig *cns.GetNetworkContainerResponse, subnetPref
 
 func getNetworkName(podName, podNs, ifName string, nwCfg *cni.NetworkConfig) (string, error) {
 	if nwCfg.MultiTenancy {
-		_, cnsNetworkConfig, _, err := getContainerNetworkConfiguration(nwCfg, "", podName, podNs, ifName)
+		_, cnsNetworkConfig, _, err := getContainerNetworkConfiguration(nwCfg, podName, podNs, ifName)
```
Comment (ninzavivek) [inline cni/network/network_windows.go]: As discussed, if we don't handle an operation ona  given platform we should explicitly throw an error, operation is not support rather silently not doing anything.

File: cni/network/network_windows.go [hunk 16]
```diff
@@ -111,7 +111,7 @@ func updateSubnetPrefix(cnsNwConfig *cns.GetNetworkContainerResponse, subnetPref
 
 func getNetworkName(podName, podNs, ifName string, nwCfg *cni.NetworkConfig) (string, error) {
```
Comment (ninzavivek) [inline cni/network/network_windows.go]: As discussed, if we don't handle an operation on a given platform we should explicitly throw an error, operation is not support rather silently not doing anything.

File: cns/restserver/restserver.go (hunk 12 above)
Comment (jaer-tsun) [reply cns/restserver/restserver.go]: done

File: cns/restserver/restserver.go (hunk 14 above)
Comment (jaer-tsun) [reply cns/restserver/restserver.go]: done

File: cns/restserver/restserver.go (hunk 13 above)
Comment (jaer-tsun) [reply cns/restserver/restserver.go]: done
